lib/ex_unit/lib/ex_unit/assertions.ex

```diff
@@ -112,14 +112,14 @@ defmodule ExUnit.Assertions do
     check =
       suppress_warning(
         quote do
-          case right do
-            x when x in [nil, false] ->
+          case truthy?(right) do
```

Comment:
Suppress warnings should have been enough. Does it work if we change this line:

https://github.com/elixir-lang/elixir/blob/8a2f3783caf695b68d443244937511ec924177f2/lib/ex_unit/lib/ex_unit/assertions.ex#L731

To:

    {name, [generated: true] ++ meta, [expr, [do: clauses]]}

?

Reply:
I tried it as well, and [as reported](https://github.com/elixir-lang/elixir/issues/11869#issuecomment-1160647565=) by @Ljzn it doesn't work.
I also tried to add `generated: true` to the nested `:when` and `:in` without any success either.
I'll keep looking into why dialyzer doesn't seem to understand `generated: true` in this context 🤔

Reply:
If it doesn't work, I would rather close this as won't fix, because in this case I think it is a Dialyzer bug. We might want to keep the `unless` one though!

Reply:
OK I now understand what is going on, the bug was not in dialyzer after all 🙂.

The following fixes it:
```elixir
  check =
      quote generated: true do
        case right do
          # we cannot use in/2 sugar here because we need generated: true
          x when x == nil or x == false ->
             ...
```

- we need to use `generated: true` in the quote instead of `suppress_warnings`, since it needs to be added to the guard as well
- we can't use the `in/2` macro which doesn't pass on `generated: true` to the generated `orelse` code. This could be improved, but it is also probably the case for other macros such as `if` etc.

 If this approach sounds OK, I will submit a fix once I've got the dialyzer tests ready.

Reply:
> We should indeed also consider making expansion of generated macros also generated.

Should I open an issue for this to start the discussion? I'm interested in giving it a try if this is something we want to move forward with, although this sounds quite difficult.

Regarding the strategy, would it be something like:
- adding a `generated` field to `Macro.Env`
- populate it in `__CALLER__` when the macro has `generated: true`
- in relevant macros, replace `quote do` by `quote generated: __CALLER__.generated do` (or should this be the default?)